app/models/ad.rb

```diff
@@ -1,14 +1,27 @@
 class Ad < ApplicationRecord
   belongs_to :peer, optional: true # set if the ad comes from a peer
   has_many :ad_peers, dependent: :destroy # who has been told about this ad
-  has_many :webhook_sends, as: :resource, dependent: :destroy, class_name: 'Webhook::Send'
-  has_many :webhook_receipts, as: :resource, dependent: :destroy, class_name: 'Webhook::Receipt'
+  has_many :webhook_sends, as: :resource, class_name: 'Webhook::Send'
+  has_many :webhook_receipts, as: :resource, class_name: 'Webhook::Receipt'
 
+  after_initialize :generate_uuid
   after_commit :propagate_to_peers
 
+  validates_presence_of :uuid
```

Comment:
should validate_uniqueness_of as well.

which implies the 'what if 2 peers are telling me about the same ad update' needs better coverage as well.